external/fv3fit/tests/reservoir/test_transformer.py

```diff
@@ -23,37 +23,134 @@ def test_DoNothingAutoencoder(nz, nvars):
     "nt, nx, ny, nz, nvars",
     [(20, 4, 4, 3, 2), (None, 2, 2, 1, 1), (None, 2, 2, None, 1)],
 )
-def test_encode_columns(nt, nx, ny, nz, nvars):
+def test_base_encode_txyz(nt, nx, ny, nz, nvars):
     shape = tuple([y for y in [nt, nx, ny, nz] if y is not None])
     expected_shape = (*shape[:-1], nz * nvars) if nz is not None else shape
     transformer = DoNothingAutoencoder([nz for var in range(nvars)])
     data_arrs = [np.random.rand(*shape) for var in range(nvars)]
 
-    encoded = encode_columns(data_arrs, transformer=transformer)
+    encoded = transformer.encode_txyz(data_arrs)
 
     assert encoded.shape == expected_shape
 
 
 @pytest.mark.parametrize(
     "nx, ny, nz, nvars", [(4, 4, 3, 2), (2, 2, 1, 1), (2, 2, 1, 2)]
 )
-def test_decode_columns(nx, ny, nz, nvars):
-    expected_shapes = [
-        tuple([y for y in [nx, ny, nz] if y is not None]) for var in range(nvars)
-    ]
-
-    encoded_input_shape = (
-        (*expected_shapes[0][:-1], nz * nvars) if nz is not None else expected_shapes[0]
-    )
+def test_base_decode_txyz(nx, ny, nz, nvars):
+    expected_shapes = [(nx, ny, nz) for var in range(nvars)]
+
     transformer = DoNothingAutoencoder([nz for var in range(nvars)])
 
     # need to call encode before decode
     data_arrs = [np.random.rand(*shape) for shape in expected_shapes]
-    transformer.encode(data_arrs)
+    transformer.encode_txyz(data_arrs)
 
-    encoded_input = np.random.rand(*encoded_input_shape)
-    decoded = decode_columns(encoded_input, transformer=transformer)
+    encoded_input = np.random.rand(nx, ny, transformer.n_latent_dims)
+    decoded = transformer.decode_txyz(encoded_input)
 
     assert len(expected_shapes) == len(decoded)
     for expected_shape, decoded_output in zip(expected_shapes, decoded):
         assert expected_shape == decoded_output.shape
+
+
+def _get_sample_xyz_data(nx, ny, nz):
+    xyz_base = np.ones((1, nx, ny, nz))
+    a = np.concatenate([-10 * xyz_base, 10 * xyz_base], axis=0)  # mean = 0, std = 10
+    b = np.concatenate([xyz_base, 2 * xyz_base], axis=0)  # mean = 1, std = 1
+    return [a, b]
+
+
+# test encode w/ specified arrays
+def test_scale_per_feature_concat_z_transform():
+    nx, ny, nz = 3, 4, 1
+    a, b = _get_sample_xyz_data(nx, ny, nz)
+
+    xyz_like = np.ones((1, nx, ny, nz))
+    normalized = np.concatenate([-1 * xyz_like, 1 * xyz_like], axis=0)
+    normalized_and_stacked = np.concatenate([normalized, normalized], axis=-1)
+
+    transformer = build_scale_spatial_concat_z_transformer([a, b])
+    encoded = transformer.encode_txyz([a, b])
+
+    # test that it normalizes
+    nsamples, nvars = 2, 2
+    assert encoded.shape == (nsamples, 3, 4, 1 * nvars)
+    np.testing.assert_allclose(encoded, normalized_and_stacked, rtol=1e-6)
+
+    # test round trip
+    decoded = transformer.decode_txyz(encoded)
+    np.testing.assert_allclose(decoded, [a, b], rtol=1e-6)
+
```

Comment:
Just to cover bases, could you add a test for the new transformer decoding an encoded input that has no time dim? I think this is how it is used in the prognostic run.

Reply:
Added